libs/ml/src/model/model.cpp

```diff
@@ -0,0 +1,414 @@
+ * @tparam TensorType
+ * @return
+ */
+template <typename TensorType>
+void Model<TensorType>::Train()
+{
+  model_config_.subset_size = fetch::ml::optimisers::SIZE_NOT_SET;
+  DataType _;
+  Model<TensorType>::TrainImplementation(_);
+}
+
+/**
+ * An interface to train that doesn't report loss
+ * @tparam TensorType
+ * @param n_rounds
+ * @return
+ */
+template <typename TensorType>
+void Model<TensorType>::Train(SizeType n_rounds)
+{
+  DataType _;
+  Model<TensorType>::Train(n_rounds, _);
+}
+
+template <typename TensorType>
+void Model<TensorType>::Train(SizeType n_rounds, DataType &loss)
+{
+  TrainImplementation(loss, n_rounds);
+}
+
+template <typename TensorType>
+void Model<TensorType>::Test(DataType &test_loss)
+{
+  if (!compiled_)
+  {
+    throw ml::exceptions::InvalidMode("must compile model before testing");
+  }
+  if (!DataLoaderIsSet())
+  {
+    throw ml::exceptions::InvalidMode("must set data before testing");
+  }
+
+  dataloader_ptr_->SetMode(dataloaders::DataLoaderMode::TEST);
+
+  SizeType test_set_size = dataloader_ptr_->Size();
+
+  dataloader_ptr_->Reset();
+  bool is_done_set;
+  auto test_pair = dataloader_ptr_->PrepareBatch(test_set_size, is_done_set);
+
+  this->graph_ptr_->SetInput(label_, test_pair.first);
+  this->graph_ptr_->SetInput(input_, test_pair.second.at(0));
+  test_loss = *(this->graph_ptr_->Evaluate(error_).begin());
+}
+
+template <typename TensorType>
+void Model<TensorType>::Predict(TensorType &input, TensorType &output)
+{
+  if (!compiled_)
+  {
+    throw ml::exceptions::InvalidMode("must compile model before predicting");
+  }
+
+  this->graph_ptr_->SetInput(input_, input);
+  output = this->graph_ptr_->Evaluate(output_);
+}
+
+template <typename TensorType>
+typename Model<TensorType>::DataVectorType Model<TensorType>::Evaluate(
+    dataloaders::DataLoaderMode dl_mode, SizeType batch_size)
+{
+  if (!compiled_)
+  {
+    throw ml::exceptions::InvalidMode("must compile model before evaluating");
+  }
+  if (!DataLoaderIsSet())
+  {
+    throw ml::exceptions::InvalidMode("must set data before evaluating");
+  }
+
+  dataloader_ptr_->SetMode(dl_mode);
+  dataloader_ptr_->SetRandomMode(false);
+  bool is_done_set;
+  if (batch_size == 0)
+  {
+    batch_size = dataloader_ptr_->Size();
+  }
+  auto evaluate_pair = dataloader_ptr_->PrepareBatch(batch_size, is_done_set);
+
+  this->graph_ptr_->SetInput(label_, evaluate_pair.first);
+  this->graph_ptr_->SetInput(input_, evaluate_pair.second.at(0));
+
+  DataVectorType ret;
+  // call evaluate on the graph and store the loss in ret
+  ret.emplace_back(*(this->graph_ptr_->Evaluate(error_).cbegin()));
+
+  // push further metrics into ret - due to graph caching these subsequent evaluate calls are cheap
+  for (auto const &met : metrics_)
+  {
+    ret.emplace_back(*(this->graph_ptr_->Evaluate(met).cbegin()));
+  }
+  return ret;
+}
+
+template <typename TensorType>
+template <typename... Params>
+void Model<TensorType>::SetData(Params... params)
```

Comment:
as before - you shouldn't move this type of implementation into source because you can't explicitly instantiate every possible set of types that Params could mean. move it back into header.

Reply:
Done